server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2408,18 +2409,63 @@ private void deleteSnapshotsFromRepository(
                 // delete now, we can avoid reaching to the repository and can complete the deletion.
                 // TODO we should complete the deletion and resolve the listeners of SnapshotDeletionsInProgress with no snapshot sooner,
                 // that would save some cluster state updates.
-                removeSnapshotDeletionFromClusterState(deleteEntry, null, repositoryData);
+                removeSnapshotDeletionFromClusterState(
+                    deleteEntry,
+                    repositoryData,
+                    listeners -> completeListenersIgnoringException(listeners, null)
+                );
                 return;
             }
             repositoriesService.repository(deleteEntry.repository())
                 .deleteSnapshots(
                     snapshotIds,
                     repositoryData.getGenId(),
                     minCompatibleVersion(minNodeVersion, repositoryData, snapshotIds),
-                    ActionListener.wrap(updatedRepoData -> {
-                        logger.info("snapshots {} deleted", snapshotIds);
-                        removeSnapshotDeletionFromClusterState(deleteEntry, null, updatedRepoData);
-                    }, ex -> removeSnapshotDeletionFromClusterState(deleteEntry, ex, repositoryData))
+                    new SnapshotDeleteListener() {
+
+                        private final ListenableFuture<Void> doneFuture = new ListenableFuture<>();
+
+                        @Override
+                        public void onDone() {
+                            logger.info("snapshots {} deleted", snapshotIds);
+                            doneFuture.onResponse(null);
+                        }
+
+                        @Override
+                        public void onMetaUpdated(RepositoryData updatedRepoData) {
+                            removeSnapshotDeletionFromClusterState(
```

Comment:
It's true that this leads to a little more leaking on the shard level statistically speaking but I'm not so worried about it. If you think about it, it's quite unlikely for an index to run into a a partially failed delete and then not be touched again for a long time. The only scenario in which this would realistically happen is when a delete fails for the last regular backup of a shard/index that has moved to frozen and thus still has a snapshot around.
In that case though, the contents of the frozen-tier backing snapshot are probably the same as those of the last backup anyway so idk.
Master fail-overs + unlikely and I can't really see a common case where we'd have a massive diff between snapshots in a shard and then never touch it again => seems not too important?
Also note, that this brings us back to how things worked pre-7.6 in a sense (where we didn't track repository generation in the CS and thus the second delete after fail-over wouldn't do anything) and in my recollection, we hardly had any blobs leak past 7.4 back then (we did collect some data on cloud about that).

Reply:
I share David's concerns but on the other hand I think this change has benefits that outweigh the (apparently unlikely) inconvenience of leaking blobs, so I'm +1 on this as long as we keep an eye on this on Cloud